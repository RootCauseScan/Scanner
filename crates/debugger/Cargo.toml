[package]
name = "debugger"
version = "1.0.0"
edition = "2021"
description = "Interactive debugger for inspecting RootCause rules and IR"
readme = "README.md"
license.workspace = true
repository = "https://github.com/rootcausescan/scanner"
homepage = "https://github.com/rootcausescan/scanner"
keywords = ["security", "sast", "static-analysis"]
categories = ["development-tools", "code-analysis"]

[[bin]]
name = "debugger"
path = "src/main.rs"

[lib]
path = "src/lib.rs"

[dependencies]
anyhow = "1"
clap = { version = "4", features = ["derive"] }
ir = { path = "../ir" }
loader = { path = "../loader" }
parsers = { path = "../parsers" }
serde_json = { workspace = true }
serde = { workspace = true }
engine = { path = "../engine" }

[dev-dependencies]
insta = "1"

