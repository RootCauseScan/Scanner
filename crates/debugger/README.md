# debugger

Lightweight tool for inspecting the intermediate representation and compiled rules.

## Examples

Get the AST of a file in DOT format:

```bash
cargo run -p debugger -- ir examples/fixtures/python/py.no-eval/bad.py --kind ast --format dot
```

Get the CFG in JSON format:

```bash
cargo run -p debugger -- ir examples/fixtures/python/py.no-eval/bad.py --kind cfg --format json
```

Show a compiled rule in text format:

```bash
cargo run -p debugger -- rule examples/rules/python/py.no-exec.yaml --format text
```

## Contributing

See [CONTRIBUTING.md](../CONTRIBUTING.md) for guidelines.
