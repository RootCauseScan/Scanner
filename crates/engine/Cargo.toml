[package]
name = "engine"
version = "1.0.0"
edition = "2021"
description = "Parallel rule evaluation engine for RootCause"
readme = "README.md"
license.workspace = true
repository = "https://github.com/rootcausescan/scanner"
homepage = "https://github.com/rootcausescan/scanner"
keywords = ["security", "sast", "static-analysis"]
categories = ["development-tools", "code-analysis"]

[dependencies]
rayon = { workspace = true }
regex = { workspace = true }
serde = { workspace = true, features = ["derive"] }
serde_json = { workspace = true }
loader = { path = "../loader" }
ir = { path = "../ir" }
rego_wasm = { path = "../rego_wasm" }
tokio = { version = "1", features = ["rt-multi-thread", "time"] }
anyhow = "1"
tracing = { workspace = true }
blake3 = "1"
parsers = { path = "../parsers" }
shlex = { workspace = true }
plugin_core = { path = "../plugin_core" }

[dev-dependencies]
parsers = { path = "../parsers" }
tempfile = "3"
tracing-subscriber = { workspace = true }
base64 = { workspace = true }
