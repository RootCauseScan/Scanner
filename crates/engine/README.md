# engine

Analysis engine that evaluates rules over the intermediate representation with parallel execution and cache.

## Build

```bash
cargo build -p engine
```

## Test

```bash
cargo test -p engine
```

## Documentation

- [Architecture details](../../documentation/architecture/crates/engine.md)

## Contributing

See [CONTRIBUTING.md](../CONTRIBUTING.md) for guidelines.
