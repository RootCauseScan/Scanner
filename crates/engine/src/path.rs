use regex::Regex;
use std::collections::{HashMap, VecDeque};
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::{Mutex, OnceLock, RwLock};

pub static CANONICAL_PATHS: OnceLock<RwLock<HashMap<PathBuf, PathBuf>>> = OnceLock::new();
pub static CANONICAL_STATS: OnceLock<CacheStats> = OnceLock::new();
pub static CANONICAL_ORDER: OnceLock<RwLock<VecDeque<PathBuf>>> = OnceLock::new();
static PATH_REGEX_CACHE: OnceLock<Mutex<PathRegexCache>> = OnceLock::new();

#[cfg(test)]
pub const CANONICAL_CACHE_CAPACITY: usize = 3;
#[cfg(not(test))]
pub const CANONICAL_CACHE_CAPACITY: usize = 1024;

#[cfg(test)]
pub const PATH_REGEX_CACHE_CAPACITY: usize = 3;
#[cfg(not(test))]
pub const PATH_REGEX_CACHE_CAPACITY: usize = 1024;

#[derive(Default)]
pub struct CacheStats {
    hits: AtomicUsize,
    misses: AtomicUsize,
}

struct PathRegexCache {
    map: HashMap<String, Regex>,
    order: VecDeque<String>,
    capacity: usize,
}

impl PathRegexCache {
    fn new(capacity: usize) -> Self {
        Self {
            map: HashMap::new(),
            order: VecDeque::new(),
            capacity,
        }
    }

    fn get(&mut self, key: &str) -> Option<&Regex> {
        if let Some(pos) = self.order.iter().position(|k| k == key) {
            let k = self.order.remove(pos).expect("order index must exist");
            self.order.push_back(k);
            return self.map.get(key);
        }
        None
    }

    fn insert(&mut self, key: String, value: Regex) {
        if self.map.contains_key(&key) {
            if let Some(pos) = self.order.iter().position(|k| k == &key) {
                self.order.remove(pos);
            }
        }
        self.order.push_back(key.clone());
        self.map.insert(key, value);
        if self.order.len() > self.capacity {
            if let Some(oldest) = self.order.pop_front() {
                self.map.remove(&oldest);
            }
        }
    }

    #[cfg(test)]
    fn clear(&mut self) {
        self.map.clear();
        self.order.clear();
    }

    #[cfg(test)]
    fn len(&self) -> usize {
        self.map.len()
    }

    #[cfg(test)]
    fn contains(&self, key: &str) -> bool {
        self.map.contains_key(key)
    }
}

pub fn canonicalize_path<P: AsRef<Path>>(path: P) -> PathBuf {
    let path_ref = path.as_ref();
    let cache = CANONICAL_PATHS.get_or_init(|| RwLock::new(HashMap::new()));
    let order = CANONICAL_ORDER.get_or_init(|| RwLock::new(VecDeque::new()));
    let stats = CANONICAL_STATS.get_or_init(Default::default);
    if let Some(cached) = cache
        .read()
        .unwrap_or_else(|e| e.into_inner())
        .get(path_ref)
        .cloned()
    {
        stats.hits.fetch_add(1, Ordering::Relaxed);
        let mut ord = order.write().unwrap_or_else(|e| e.into_inner());
        if let Some(pos) = ord.iter().position(|p| p == path_ref) {
            ord.remove(pos);
        }
        ord.push_back(path_ref.to_path_buf());
        return cached;
    }
    stats.misses.fetch_add(1, Ordering::Relaxed);
    let canonical = fs::canonicalize(path_ref).unwrap_or_else(|_| {
        #[cfg(windows)]
        {
            PathBuf::from(path_ref.to_string_lossy().replace('\\', "/"))
        }
        #[cfg(not(windows))]
        {
            path_ref.to_path_buf()
        }
    });
    let mut map = cache.write().unwrap_or_else(|e| e.into_inner());
    let mut ord = order.write().unwrap_or_else(|e| e.into_inner());
    map.insert(path_ref.to_path_buf(), canonical.clone());
    ord.push_back(path_ref.to_path_buf());
    if ord.len() > CANONICAL_CACHE_CAPACITY {
        if let Some(oldest) = ord.pop_front() {
            map.remove(&oldest);
        }
    }
    canonical
}

pub(crate) fn cache_stats() -> (usize, usize) {
    let stats = CANONICAL_STATS.get_or_init(Default::default);
    (
        stats.hits.load(Ordering::Relaxed),
        stats.misses.load(Ordering::Relaxed),
    )
}

#[cfg(test)]
pub fn reset_canonical_cache() {
    if let Some(map) = CANONICAL_PATHS.get() {
        map.write().unwrap_or_else(|e| e.into_inner()).clear();
    }
    if let Some(ord) = CANONICAL_ORDER.get() {
        ord.write().unwrap_or_else(|e| e.into_inner()).clear();
    }
    if let Some(stats) = CANONICAL_STATS.get() {
        stats.hits.store(0, Ordering::Relaxed);
        stats.misses.store(0, Ordering::Relaxed);
    }
}

#[cfg(test)]
pub fn canonical_cache_stats() -> (usize, usize) {
    cache_stats()
}

#[cfg(test)]
pub fn reset_path_regex_cache() {
    if let Some(cache) = PATH_REGEX_CACHE.get() {
        cache.lock().unwrap().clear();
    }
}

#[cfg(test)]
pub fn path_regex_cache_size() -> usize {
    PATH_REGEX_CACHE
        .get()
        .map(|c| c.lock().unwrap().len())
        .unwrap_or(0)
}

#[cfg(test)]
pub fn path_regex_cache_contains(pat: &str) -> bool {
    PATH_REGEX_CACHE
        .get()
        .map(|c| c.lock().unwrap().contains(pat))
        .unwrap_or(false)
}

pub fn path_matches(pattern: &str, candidate: &str) -> bool {
    if !pattern
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || "/._-".contains(c) || c == '*')
    {
        return false;
    }
    let cache =
        PATH_REGEX_CACHE.get_or_init(|| Mutex::new(PathRegexCache::new(PATH_REGEX_CACHE_CAPACITY)));
    let mut cache = cache.lock().expect("path regex cache lock poisoned");
    if let Some(rx) = cache.get(pattern) {
        return rx.is_match(candidate);
    }
    let mut re = String::from("^");
    for ch in pattern.chars() {
        if ch == '*' {
            re.push_str(".*");
        } else {
            re.push_str(&regex::escape(&ch.to_string()));
        }
    }
    re.push('$');
    match Regex::new(&re) {
        Ok(rx) => {
            let is_match = rx.is_match(candidate);
            cache.insert(pattern.to_string(), rx);
            is_match
        }
        Err(_) => false,
    }
}
