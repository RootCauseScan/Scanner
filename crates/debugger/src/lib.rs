use engine::debug::{DebugEvent, DebugSink};
use serde::Serialize;
use std::path::Path;
use std::sync::{Arc, Mutex};

#[derive(Clone, Copy)]
pub enum EventFormat {
    Text,
    Json,
}

#[derive(Clone)]
pub struct FormatterSink {
    format: EventFormat,
    lines: Arc<Mutex<Vec<String>>>,
}

impl FormatterSink {
    pub fn new(format: EventFormat) -> Self {
        Self {
            format,
            lines: Arc::new(Mutex::new(Vec::new())),
        }
    }

    pub fn output(&self) -> String {
        self.lines
            .lock()
            .expect("debug sink lock poisoned")
            .join("\n")
    }
}

#[derive(Serialize, Debug)]
struct SerEvent {
    kind: &'static str,
    path: Option<String>,
    file: Option<String>,
    rule_id: Option<String>,
    matched: Option<bool>,
}

fn name(p: &Path) -> String {
    p.file_name()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_default()
}

impl From<&DebugEvent> for SerEvent {
    fn from(e: &DebugEvent) -> Self {
        match e {
            DebugEvent::ParseStart { path } => SerEvent {
                kind: "ParseStart",
                path: Some(name(path)),
                file: None,
                rule_id: None,
                matched: None,
            },
            DebugEvent::ParseEnd { path } => SerEvent {
                kind: "ParseEnd",
                path: Some(name(path)),
                file: None,
                rule_id: None,
                matched: None,
            },
            DebugEvent::RuleCompiled { id } => SerEvent {
                kind: "RuleCompiled",
                path: None,
                file: None,
                rule_id: Some(id.clone()),
                matched: None,
            },
            DebugEvent::MatchAttempt { rule_id, file } => SerEvent {
                kind: "MatchAttempt",
                path: None,
                file: Some(name(file)),
                rule_id: Some(rule_id.clone()),
                matched: None,
            },
            DebugEvent::MatchResult {
                rule_id,
                file,
                matched,
            } => SerEvent {
                kind: "MatchResult",
                path: None,
                file: Some(name(file)),
                rule_id: Some(rule_id.clone()),
                matched: Some(*matched),
            },
        }
    }
}

impl DebugSink for FormatterSink {
    fn event(&self, event: DebugEvent) {
        let ser = SerEvent::from(&event);
        let line = match self.format {
            EventFormat::Text => format!("{ser:?}"),
            EventFormat::Json => serde_json::to_string(&ser).expect("serialize debug event"),
        };
        self.lines
            .lock()
            .expect("debug sink lock poisoned")
            .push(line);
    }
}
