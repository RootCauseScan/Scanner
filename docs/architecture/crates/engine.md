---
id: engine
title: Engine
description: Analysis engine that evaluates rules on the intermediate representation
sidebar_position: 2
---

# engine

## Responsibility
Evaluates rules on the intermediate representation of files, handling parallel execution, caching and optional WASM support.

## Key functions and structures
- `analyze_files_cached`: analyzes files applying hash-based caching.
- `build_cfg`: builds control flow graphs for flow analysis.
- `Finding`: individual result with severity, location and message.

## Usage example
```rust
use engine::{analyze_files_cached, EngineConfig};
use loader::load_rules;
use parsers::parse_file;
use std::path::Path;

let rules = load_rules(Path::new("rules")).unwrap();
let file = parse_file(Path::new("a.yaml"), None, None).unwrap().unwrap();
let findings = analyze_files_cached(&[file], &rules, Path::new("cache.json"), &EngineConfig::default(), None);
assert!(findings.is_empty());
```
