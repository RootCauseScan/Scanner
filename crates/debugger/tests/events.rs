use debugger::{EventFormat, FormatterSink};
use engine::{
    analyze_file, load_rules_with_events, parse_file_with_events, reset_rule_cache, set_debug_sink,
};
use std::path::Path;
use std::sync::Mutex;

static TEST_LOCK: Mutex<()> = Mutex::new(());

#[test]
fn debug_events_text() {
    let _guard = TEST_LOCK.lock().unwrap();
    reset_rule_cache();
    let sink = FormatterSink::new(EventFormat::Text);
    set_debug_sink(Some(Box::new(sink.clone())));
    let rules =
        load_rules_with_events(Path::new("../../examples/rules/python/py.no-exec.yaml")).unwrap();
    let ir = parse_file_with_events(
        Path::new("../../examples/fixtures/python/py.no-exec/bad.py"),
        None,
        None,
    )
    .unwrap()
    .unwrap();
    analyze_file(&ir, &rules);
    set_debug_sink(None);
    insta::assert_snapshot!(sink.output());
}

#[test]
fn debug_events_json() {
    let _guard = TEST_LOCK.lock().unwrap();
    reset_rule_cache();
    let sink = FormatterSink::new(EventFormat::Json);
    set_debug_sink(Some(Box::new(sink.clone())));
    let rules =
        load_rules_with_events(Path::new("../../examples/rules/python/py.no-exec.yaml")).unwrap();
    let ir = parse_file_with_events(
        Path::new("../../examples/fixtures/python/py.no-exec/bad.py"),
        None,
        None,
    )
    .unwrap()
    .unwrap();
    analyze_file(&ir, &rules);
    set_debug_sink(None);
    insta::assert_snapshot!(sink.output());
}
