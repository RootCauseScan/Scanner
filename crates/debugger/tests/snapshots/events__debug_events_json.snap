---
source: crates/debugger/tests/events.rs
expression: sink.output()
---
{"kind":"RuleCompiled","path":null,"file":null,"rule_id":"py.no-exec","matched":null}
{"kind":"ParseStart","path":"bad.py","file":null,"rule_id":null,"matched":null}
{"kind":"ParseEnd","path":"bad.py","file":null,"rule_id":null,"matched":null}
{"kind":"MatchAttempt","path":null,"file":"bad.py","rule_id":"py.no-exec","matched":null}
{"kind":"MatchResult","path":null,"file":"bad.py","rule_id":"py.no-exec","matched":true}
